sleepy/types.py

```diff
@@ -512,21 +513,29 @@ def _make_di_type(self, context: CodegenContext) -> ir.DIValue:
       })
 
   def is_realizable(self) -> bool:
-    return len(self.possible_types) > 0
+    return len(self.possible_types) > 0 and all(possible_type.is_realizable() for possible_type in self.possible_types)
 
   def replace_types(self, replacements: Dict[Type, Type]) -> Type:
     if self in replacements:
       return replacements[self]
     if len(self.possible_types) == 0:
       return self
-    new_possible_types = [possible_type.replace_types(replacements) for possible_type in self.possible_types]
-    new_possible_type_nums = self.possible_type_nums.copy()
-    duplicate_indices = [
-      index for index, possible_type in enumerate(new_possible_types)
-      if possible_type in new_possible_types[:index]]
-    for duplicate_index in reversed(duplicate_indices):
-      del new_possible_types[duplicate_index]
-      del new_possible_type_nums[duplicate_index]
+    replaced_possible_types = [possible_type.replace_types(replacements) for possible_type in self.possible_types]
+    new_possible_types: List[Type] = []
+    new_possible_type_nums: List[int] = []
+    for replaced_type in replaced_possible_types:
+      possible_replaced_types = (
+        replaced_type.possible_types if isinstance(replaced_type, UnionType) else [replaced_type])
+      for possible_type in possible_replaced_types:
+        if possible_type in new_possible_types:
+          continue
+        if possible_type in self.possible_types:  # keep type nums as before
+          new_possible_type_num = self.get_variant_num(possible_type)
+        else:
+          new_possible_type_num = max(new_possible_type_nums, default=-1) + 1
+        new_possible_types.append(possible_type)
+        new_possible_type_nums.append(new_possible_type_num)
+    assert len(new_possible_types) == len(new_possible_type_nums)
```

Comment:
If I understand correctly, we have 3 steps here: replacement for each possible type, one level flattening, calculation of indices. Should make this clearer in code or if necessary with comments